### ISSUE TITLE
memory leak detected by CRAN

### ISSUE THREAD

**Issue by pachadotdev** [litalbarkai/redatam-converter#11]
Hi

My code had this memory leak on [clang-ASAN](https://www.stats.ox.ac.uk/pub/bdr/memtests/clang-ASAN/redatam) [gcc-UBSAN](https://www.stats.ox.ac.uk/pub/bdr/memtests/gcc-UBSAN/redatam)

the errors are these and appear when running `devtools::test()` in https://github.com/pachadotdev/open-redatam/tree/clang-asan

```
> devtools::test()
ℹ Testing redatam
✔ | F W  S  OK | Context
⠏ |          0 | galapagos                                                                                  Opening dictionary file...
=================================================================
==64683==ERROR: AddressSanitizer: alloc-dealloc-mismatch (operator new vs free) on 0x50600003e660
    #0 0x59a8a24a4356 in free (/home/pacha/github/open-redatam/rpkg/dev/R-devel/bin/exec/R+0xcb356) (BuildId: e68667ae741dfff9d511b6cf00e1bbc0361b9d69)
    #1 0x7006faf9721d in std::invalid_argument::~invalid_argument() (/usr/lib/x86_64-linux-gnu/libc++abi.so.1+0x2421d) (BuildId: 002aba7391c884e78114f6cb82186797d9edbf4b)
    #2 0x7006faf99593 in __cxa_end_catch (/usr/lib/x86_64-linux-gnu/libc++abi.so.1+0x26593) (BuildId: 002aba7391c884e78114f6cb82186797d9edbf4b)
    #3 0x7006f0ff9986 in RedatamLib::ByteArrayReader::TryReadStr(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>*, bool) /home/pacha/github/open-redatam/rpkg/src/redatamlib/readers/ByteArrayReader.cpp:123:3
    #4 0x7006f10219d6 in RedatamLib::FuzzyEntityParser::TryGetEntity() /home/pacha/github/open-redatam/rpkg/src/redatamlib/readers/FuzzyEntityParser.cpp:45:17
    #5 0x7006f1020e09 in RedatamLib::FuzzyEntityParser::ParseEntities() /home/pacha/github/open-redatam/rpkg/src/redatamlib/readers/FuzzyEntityParser.cpp:19:14
    #6 0x7006f10d64f6 in RedatamLib::RedatamDatabase::OpenDictionary(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) /home/pacha/github/open-redatam/rpkg/src/redatamlib/entities/RedatamDatabase.cpp:33:25
    #7 0x7006f10d6296 in RedatamLib::RedatamDatabase::RedatamDatabase(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) /home/pacha/github/open-redatam/rpkg/src/redatamlib/entities/RedatamDatabase.cpp:20:3
    #8 0x7006f10eae56 in export_redatam_to_list_(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) /home/pacha/github/open-redatam/rpkg/src/main.cpp:12:33
    #9 0x7006f10eca2c in _redatam_export_redatam_to_list_ /home/pacha/github/open-redatam/rpkg/src/cpp11.cpp:12:27
    #10 0x7006fb5fa14a in R_doDotCall /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/dotcode.c:754:11
    #11 0x7006fb646c62 in do_dotcall /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/dotcode.c:1437:11
    #12 0x7006fb74ca7b in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1260:9
    #13 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #14 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #15 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #16 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #17 0x7006fb74cc5f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #18 0x7006fb74cc5f in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1280:12
    #19 0x7006fb765275 in do_set /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3571:8
    #20 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #21 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #22 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #23 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #24 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #25 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #26 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #27 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #28 0x7006fb74cc5f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #29 0x7006fb74cc5f in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1280:12
    #30 0x7006fb765275 in do_set /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3571:8
    #31 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #32 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #33 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #34 0x7006fb76ac62 in do_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3945:13
    #35 0x7006fb799337 in bcEval_loop /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:8119:14
    #36 0x7006fb74dba4 in bcEval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:7502:16
    #37 0x7006fb74c2df in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1167:8
    #38 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #39 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #40 0x7006fb74cc5f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #41 0x7006fb74cc5f in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1280:12
    #42 0x7006fb76b2ea in do_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3963:9
    #43 0x7006fb799337 in bcEval_loop /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:8119:14
    #44 0x7006fb74dba4 in bcEval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:7502:16
    #45 0x7006fb74c2df in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1167:8
    #46 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #47 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #48 0x7006fb75442f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #49 0x7006fb75442f in R_forceAndCall /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2460:8
    #50 0x7006fb4072e5 in do_lapply /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/apply.c:75:8
    #51 0x7006fb8e0d12 in do_internal /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/names.c:1410:11
    #52 0x7006fb78a7cc in bcEval_loop /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:8139:15
    #53 0x7006fb74dba4 in bcEval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:7502:16
    #54 0x7006fb74c2df in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1167:8
    #55 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #56 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #57 0x7006fb74cc5f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #58 0x7006fb74cc5f in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1280:12
    #59 0x7006fb891677 in Rf_ReplIteration /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/main.c:265:2
    #60 0x7006fb894210 in R_ReplConsole /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/main.c:317:11
    #61 0x7006fb894034 in run_Rmainloop /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/main.c:1219:5
    #62 0x7006fb8942da in Rf_mainloop /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/main.c:1226:5
    #63 0x59a8a24e6f54 in main /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/Rmain.c:29:5
    #64 0x7006fac29d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #65 0x7006fac29e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #66 0x59a8a2405354 in _start (/home/pacha/github/open-redatam/rpkg/dev/R-devel/bin/exec/R+0x2c354) (BuildId: e68667ae741dfff9d511b6cf00e1bbc0361b9d69)

0x50600003e660 is located 0 bytes inside of 54-byte region [0x50600003e660,0x50600003e696)
allocated by thread T0 here:
    #0 0x59a8a24e4b1d in operator new(unsigned long) (/home/pacha/github/open-redatam/rpkg/dev/R-devel/bin/exec/R+0x10bb1d) (BuildId: e68667ae741dfff9d511b6cf00e1bbc0361b9d69)
    #1 0x7006fb003ebf in std::logic_error::logic_error(char const*) (/usr/lib/x86_64-linux-gnu/libc++.so.1+0x4eebf) (BuildId: 78e555136344730317fc05363f9e8afdd6f83ca4)
    #2 0x7006f0ff9805 in RedatamLib::ByteArrayReader::TryReadStr(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>*, bool) /home/pacha/github/open-redatam/rpkg/src/redatamlib/readers/ByteArrayReader.cpp:114:13
    #3 0x7006f10219d6 in RedatamLib::FuzzyEntityParser::TryGetEntity() /home/pacha/github/open-redatam/rpkg/src/redatamlib/readers/FuzzyEntityParser.cpp:45:17
    #4 0x7006f1020e09 in RedatamLib::FuzzyEntityParser::ParseEntities() /home/pacha/github/open-redatam/rpkg/src/redatamlib/readers/FuzzyEntityParser.cpp:19:14
    #5 0x7006f10d64f6 in RedatamLib::RedatamDatabase::OpenDictionary(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) /home/pacha/github/open-redatam/rpkg/src/redatamlib/entities/RedatamDatabase.cpp:33:25
    #6 0x7006f10d6296 in RedatamLib::RedatamDatabase::RedatamDatabase(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) /home/pacha/github/open-redatam/rpkg/src/redatamlib/entities/RedatamDatabase.cpp:20:3
    #7 0x7006f10eae56 in export_redatam_to_list_(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) /home/pacha/github/open-redatam/rpkg/src/main.cpp:12:33
    #8 0x7006f10eca2c in _redatam_export_redatam_to_list_ /home/pacha/github/open-redatam/rpkg/src/cpp11.cpp:12:27
    #9 0x7006fb5fa14a in R_doDotCall /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/dotcode.c:754:11
    #10 0x7006fb646c62 in do_dotcall /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/dotcode.c:1437:11
    #11 0x7006fb74ca7b in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1260:9
    #12 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #13 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #14 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #15 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #16 0x7006fb74cc5f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #17 0x7006fb74cc5f in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1280:12
    #18 0x7006fb765275 in do_set /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3571:8
    #19 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #20 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #21 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #22 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #23 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10
    #24 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #25 0x7006fb756b35 in R_execClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2393:22
    #26 0x7006fb75262c in applyClosure_core /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2306:16
    #27 0x7006fb74cc5f in Rf_applyClosure /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:2328:16
    #28 0x7006fb74cc5f in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1280:12
    #29 0x7006fb765275 in do_set /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3571:8
    #30 0x7006fb74c6ab in Rf_eval /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:1232:12
    #31 0x7006fb760f12 in do_begin /home/pacha/github/open-redatam/rpkg/dev/R-devel/src/main/eval.c:3000:10

SUMMARY: AddressSanitizer: alloc-dealloc-mismatch (/home/pacha/github/open-redatam/rpkg/dev/R-devel/bin/exec/R+0xcb356) (BuildId: e68667ae741dfff9d511b6cf00e1bbc0361b9d69) in free
==64683==HINT: if you don't care about these errors you may set ASAN_OPTIONS=alloc_dealloc_mismatch=0
==64683==ABORTING
```

these errors do not appear with gcc

these appear with this clang-asan setup:

```
# install clang-19

wget https://apt.llvm.org/llvm.sh -O dev/llvm.sh
chmod +x dev/llvm.sh
sudo ./dev/llvm.sh 19
sudo apt install clang-19 lldb-19 libc++-dev libc++abi-dev

export ASAN_OPTIONS=detect_leaks=0
export RJAVA_JVM_STACK_WORKAROUND=0
export RGL_USE_NULL=true
export R_DONT_USE_TK=true

sudo ln -s /usr/bin/clang-19 /usr/bin/clang
sudo ln -s /usr/bin/clang++-19 /usr/bin/clang++

# install R-devel from source

sudo apt update
sudo apt install -y build-essential gfortran libreadline-dev libx11-dev libxt-dev \
                    libpng-dev libjpeg-dev libcairo2-dev xvfb libbz2-dev \
                    libzstd-dev liblzma-dev libcurl4-openssl-dev \
                    texinfo texlive texlive-fonts-extra \
                    texlive-fonts-recommended texlive-latex-extra \
                    libpcre2-dev libharfbuzz-dev libfribidi-dev \
                    libclang-dev llvm-dev

wget https://cran.r-project.org/src/base-prerelease/R-devel.tar.gz -O dev/R-devel.tar.gz
cd dev
tar -xzvf R-devel.tar.gz
cd R-devel

# add this to config.site
## clang-ASAN, clang-UBSAN:
CC="clang -fsanitize=address,undefined -fno-sanitize=float-divide-by-zero -fno-sanitize=alignment -fno-omit-frame-pointer"
CXX="clang++ -fsanitize=address,undefined -fno-sanitize=float-divide-by-zero -fno-sanitize=alignment -fno-omit-frame-pointer -frtti -stdlib=libc++"
CFLAGS="-g -O3 -Wall -pedantic"
FFLAGS="-g -O2 -mtune=native"
CXXFLAGS="-g -O3 -Wall -pedantic -stdlib=libc++"
MAIN_LD="clang++ -fsanitize=undefined,address -stdlib=libc++"
export ASAN_OPTIONS='detect_leaks=0:detect_odr_violation=0'
export UBSAN_OPTIONS='print_stacktrace=1'
export RGL_USE_NULL="true"
export R_DONT_USE_TK="true"

./configure \
    --prefix=/opt/R/${R_VERSION} \
    --enable-memory-profiling \
    --enable-R-shlib \
    --with-blas \
    --with-lapack

make

dev/R-devel/bin/R
# install.packages(c("knitr", "cpp11", "data.table", "stringi", "janitor", "rmarkdown", "devtools"))

dev/R-devel/bin/R
# devtools::clean_dll()
# devtools::test()
```

**Comment by pachadotdev** [comment]
fixed in 6f5fcce66b6672b4a8fcd7a3577a85fcaefec886

**Comment by pachadotdev** [comment]
R fix added to Py and cmd line tool here 4ffce65f2e1f5aaa1802007b783627eec06b136b

**Comment by pachadotdev** [comment]
full story here https://stat.ethz.ch/pipermail/r-package-devel/2024q4/011193.html
